src/complex.jl

```diff
@@ -0,0 +1,285 @@
+variable is decomposed into its real- and imaginary parts.
+
+See also [`iscomplex`](@ref iscomplex), [`isrealpart`](@ref isrealpart), [`imag`](@ref).
+"""
+Base.real(x::AbstractVariable) = x
+
+"""
+    imag(x::AbstractVariable)
+
+Return the imaginary part of a given variable if it was declared as a complex variable; else return zero.
+
+    imag(x::AbstractMonomial)
+    imag(x::AbstractVector{<:AbstractMonomial})
+    imag(x::AbstractTerm)
+    imag(x::AbstractPolynomial)
+
+Return the imaginary part of `x` by applying `imag` to all coefficients and variables; for this purpose, every complex-valued
+variable is decomposed into its real- and imaginary parts.
+
+See also [`iscomplex`](@ref iscomplex), [`isimagpart`](@ref isimagpart), [`real`](@ref).
+"""
+Base.imag(::AbstractVariable) = MA.Zero()
+
+# extend to higher-level elements. We make all those type-stable (but we need convert, as the construction method may
+# deliver simpler types than the inputs if they were deliberately casted, e.g., term to monomial)
+function iscomplex(p::AbstractPolynomialLike)
+    for v in variables(p)
+        if !iszero(maxdegree(p, v)) && iscomplex(v)
+            return true
+        end
+    end
+    return false
+end
+Base.conj(x::M) where {M<:AbstractMonomial} =
+    convert(M, reduce(*, conj(var)^exp for (var, exp) in zip(variables(x), exponents(x)); init=constantmonomial(x)))
+Base.conj(x::V) where {V<:AbstractVector{<:AbstractMonomial}} = monovec(conj.(x))
+Base.conj(x::T) where {T<:AbstractTerm} = convert(T, conj(coefficient(x)) * conj(monomial(x)))
+Base.conj(x::P) where {P<:AbstractPolynomial} = nterms(x) == 0 ? x : convert(P, sum(conj(t) for t in x))
+
+real_coefficient_type_polynomial_like(::AbstractPolynomialLike{R}) where {R<:Real} = R
+real_coefficient_type_polynomial_like(::AbstractPolynomialLike{Complex{R}}) where {R<:Real} = R
+# Real and imaginary parts are harder to realize. The real part of a monomial can easily be a polynomial.
+for fun in [:real, :imag]
+    eval(quote
+        function Base.$fun(x::Union{AbstractMonomial,AbstractTerm})
+            # We replace every complex variable by its decomposition into real and imaginary part
+            substs = [var => real(var) + 1im * imag(var) for var in variables(x) if iscomplex(var)]
+            # subs throws an error if it doesn't receive at least one substitution
+            full_version = length(substs) > 0 ? subs(x, substs...) : polynomial(x)
+            # Now everything that is imaginary can be recognized by its coefficient
+            return convert(
+                polynomialtype(full_version, real_coefficient_type_polynomial_like(full_version)),
+                mapcoefficients!($fun, full_version)
+            )
+        end
+        Base.$fun(x::AbstractVector{<:AbstractMonomial}) = map(Base.$fun, x)
+        Base.$fun(x::AbstractPolynomial{T}) where {T} = nterms(x) == 0 ?
```

Comment:
Can you explain what you mean with "directly"? Do you refer to the last function on the `AbstractPolynomial` or the monomial vector?
If you mean the last function, I guess that you refer to using the polynomial constructor that directly allows to specify a vector of terms or even a function that is applied to the coefficients. Yes, this would be a way to do it more efficiently; however, the real/imaginary part is not just applied to the coefficients, but also to the monomials, and it will expand those monomials a lot (unless we want to include a special type of monomial which corresponds to the real or imaginary part of another monomial). So this expansion would have to taken into account and combined anyway, and then I'm no longer sure if this kind of optimization will really help us a lot...

Reply:
Actually, since you do `mapcoefficients!` after, it could work, I made an inline suggestion